matplotcheck/raster.py

```diff
@@ -100,68 +97,62 @@ def assert_legend_accuracy_classified_image(
-        Then the plot image array is copied and the values are set to the
-        legend label that match the values (i.e. the element in
-        all_label_options). The same is done for the expected image array.
-        Finally those two arrays of strings are compared. Passes if they match.
         """
         # Retrieve image array
-        im_data = []
-        if self.ax.get_images():
-            im = self.ax.get_images()[0]
-            im_data, im_cmap = im.get_array(), im.get_cmap()
-        assert list(im_data), "No Image Displayed"
+        im_data = self.get_plot_image()
 
-        # Retrieve legend
-        legends = self.get_legends()
-        assert legends, "No legend displayed"
+        assert list(im_data), "No Image Displayed"
 
         # Retrieve legend entries and find which element of all_label_options
         # matches that entry
-        legend_dict = {}
-        for p in [
-            p
-            for sublist in [leg.get_patches() for leg in legends]
-            for p in sublist
-        ]:
-            label = p.get_label().lower()
-            legend_dict[p.get_facecolor()] = self._which_label(
-                label, all_label_options
-            )
 
-        # Check that each legend entry label is in one of all_label_options
-        assert len([val for val in legend_dict.values() if val]) == len(
-            all_label_options
-        ), "Incorrect legend labels"
-
-        # Create two copies of image array, one filled with the plot data class
-        # labels (im_data_labels) and the other with the expected labels
-        # (im_expected_labels)
-        im_class_dict = {}
-        for val in np.unique(im_data):
-            im_class_dict[val] = legend_dict[im_cmap(im.norm(val))]
-        im_data_labels = [
-            [im_class_dict[val] for val in row] for row in im_data.data
-        ]
-        im_expected_labels = [
-            [all_label_options[val][0] for val in row] for row in im_expected
+        labels = self.get_legend_labels()
+
+        assert len(labels) == len(all_label_options), (
+            "Number of label options provided doesn't match the number of"
+            " labels found in the image."
+        )
+
+        labels_check = [
+            self._which_label(label, all_label_options[i])
+            for i, label in enumerate(labels)
         ]
 
-        # Check that expected and actual labels match up
+        # Check that each legend entry label is in one of all_label_options
+        assert all(
+            labels_check
+        ), "Provided legend labels don't match labels found."
+
+        # Check that expected and actual arrays data match up
         assert np.array_equal(
-            im_data_labels, im_expected_labels
-        ), "Incorrect legend to data relation"
+            im_data, im_expected
+        ), "Expected image data doesn't match data in image."
 
         # IMAGE TESTS/HELPER FUNCTIONS
 
+    def get_legend_labels(self):
```

Comment:
i think we need to reorder methods - it doesn't make a lot of sense to call methods before they are defined even tho it works with classes. it would be better to organize things so the helpers come BEFORE the main method that does all of the checks.